### PR TITLE
Add a sample for one class matrix factorization

### LINKED ISSUE #1769: Need example for one-class matrix factorization
Currently we only have [one test](https://github.com/dotnet/machinelearning/blob/533e18647a77c31f8149ccf2618ccb2e354f8e62/test/Microsoft.ML.Tests/TrainerEstimators/MatrixFactorizationTests.cs#L376) as our example. We should add one into our sample project. Also, in that actual example, we need to tell user to install MatrixFactorization nuget. The ultimate scope of the new example is to cover everything user needs to do to run one-class matrix factorization.

Also, because of some settings in our test framework, our tests are not examples which can be run by copy-and-paste.

### PR DESCRIPTION
**Pull request by wschin**
Fix #1769.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            // Predicted value at row 12 and column 0 is 0.1499791 and its expected value is 0.15.
+            // Predicted value at row 13 and column 0 is 0.1499254 and its expected value is 0.15.
+            // Predicted value at row 14 and column 0 is 0.1499074 and its expected value is 0.15.
+            //
+            // Note: use the advanced options constructor to set the number of threads to 1 for a deterministic behavior.
+
+            // Two columns with highest predicted score to the 2nd row (indexed by 1). If we view row index as user ID and column as game ID,
+            // the following list contains the games recommended by the trained model. Note that sometime, you may want to exclude training
+            // data from your predicted results because those games were already purchased.
+            var topColumns = results.Where(element => element.MatrixRowIndex == 1).OrderByDescending(element => element.Score).Take(2);
```
Comment (codemzs) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >var topColumns = results.Where(element => element.MatrixRowIndex == 1).OrderByDescending(element => element.Score).Take(2); [](start = 12, length = 123)

Can we print the output for these and put in comments?

File: src/Microsoft.ML.Recommender/RecommenderCatalog.cs [hunk 2]
```diff
@@ -87,6 +87,7 @@ internal RecommendationTrainers(RecommendationCatalog catalog)
             /// <format type="text/markdown">
             /// <![CDATA[
             ///  [!code-csharp[MatrixFactorization](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/MatrixFactorizationWithOptions.cs)]
+            ///  [!code-csharp[MatrixFactorization](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs)]
```
Comment (codemzs) [inline src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: You may want to check this won't generate too much content for the user. I had 4 links for time series but after speaking with Natalie I reduced to one.

Comment (wschin) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: Matrix factorization works extremely differently with different loss functions. We must have two samples.

---
In reply to: [274225365](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274225365) [](ancestors = 274225365)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+using Microsoft.ML;
+using Microsoft.ML.Data;
+using Microsoft.ML.Trainers;
+
+namespace Samples.Dynamic.Trainers.Recommendation
+{
+    public static class OneClassMatrixFactorizationWithOptions
+    {
+        // This example shows the use of ML.NET's one-class matrix factorization module which implements
+        // Algorithm 1 in a <a href="https://www.csie.ntu.edu.tw/~cjlin/papers/one-class-mf/biased-mf-sdm-with-supp.pdf">paper</a>.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >a [](start = 26, length = 1)

"Algorithm 1 in a paper" isn't very meaningful. Can you give a longer description of what it is?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        // To run this example, it requires installation of additional nuget package
+        // <a href="https://www.nuget.org/packages/Microsoft.ML.Recommender/">Microsoft.ML.Recommender</a>.
+        public static void Example()
+        {
+            // Create a new context for ML.NET operations. It can be used for exception tracking and logging,
+            // as a catalog of available operations and as the source of randomness.
+            var mlContext = new MLContext(seed: 0);
+
+            // Get a small in-memory dataset.
+            GetOneClassMatrix(out List<MatrixElement> data, out List<MatrixElement> testData);
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >List [](start = 34, length = 4)

nit: I would use a less-specific signature than `List`.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+            // Convert the in-memory matrix into an IDataView so that ML.NET components can consume it.
+            var dataView = mlContext.Data.LoadFromEnumerable(data);
+
+            // Create a matrix factorization trainer which may consume "Value" as the training label, "MatrixColumnIndex" as the
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >may consume [](start = 59, length = 11)

takes

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            // matrix's column index, and "MatrixRowIndex" as the matrix's row index. Here nameof(...) is used to extract field
+            // names' in MatrixElement class.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: nit: Necessary?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            var options = new MatrixFactorizationTrainer.Options
+            {
+                MatrixColumnIndexColumnName = nameof(MatrixElement.MatrixColumnIndex),
+                MatrixRowIndexColumnName = nameof(MatrixElement.MatrixRowIndex),
+                LabelColumnName = nameof(MatrixElement.Value),
+                NumberOfIterations = 20,
+                NumberOfThreads = 8,
+                ApproximationRank = 32,
+                Alpha = 1,
+                // The desired of unobserved values.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >// The desired of unobserved values [](start = 16, length = 35)

Unclear what this means.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                C = 0.15,
+                // To enable one-class matrix factorization, the following line is required.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: > To enable one-class matrix factorization, the following line is required. [](start = 18, length = 74)

Suggested Rephrase: This argument enables one-class matrix factorization.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            // Train a matrix factorization model.
+            var model = pipeline.Fit(dataView);
+
+            // Apply the trained model to the test set. Notice that training is a partial 
+            var prediction = model.Transform(mlContext.Data.LoadFromEnumerable(testData));
+
+            var results = mlContext.Data.CreateEnumerable<MatrixElement>(prediction, false).ToList();
+            // Feed the test data into the model and then iterate through a few predictions.
+            foreach (var pred in results.Take(15))
+                Console.WriteLine($"Predicted value at row {pred.MatrixRowIndex - 1} and column {pred.MatrixColumnIndex - 1} is {pred.Score} and its expected value is {pred.Value}.");
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >Console.Write [](start = 16, length = 13)

Break this line; I would suggest for such long lines using old-style `{0}...{1}` string formatting so you can throw the arguments onto new lines.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            // Predicted value at row 7 and column 0 is 0.1499805 and its expected value is 0.15.
+            // Predicted value at row 8 and column 0 is 0.1500055 and its expected value is 0.15.
+            // Predicted value at row 9 and column 0 is 0.1499199 and its expected value is 0.15.
+            // Predicted value at row 10 and column 0 is 0.9873335 and its expected value is 1.
+            // Predicted value at row 11 and column 0 is 0.1499522 and its expected value is 0.15.
+            // Predicted value at row 12 and column 0 is 0.1499791 and its expected value is 0.15.
+            // Predicted value at row 13 and column 0 is 0.1499254 and its expected value is 0.15.
+            // Predicted value at row 14 and column 0 is 0.1499074 and its expected value is 0.15.
+            //
+            // Note: use the advanced options constructor to set the number of threads to 1 for a deterministic behavior.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >// Note:  [](start = 12, length = 9)

Nice touch.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+            // Two columns with highest predicted score to the 2nd row (indexed by 1). If we view row index as user ID and column as game ID,
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >Two columns with highest predicted score to the 2nd row (indexed by 1). [](start = 15, length = 71)

This sentence doesn't have any context.

Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >we [](start = 90, length = 2)

If we wanted a model to recommend video games to a user, we could view...

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            // the following list contains the games recommended by the trained model. Note that sometime, you may want to exclude training
+            // data from your predicted results because those games were already purchased.
```
Comment (rogancarr) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: >those [](start = 56, length = 5)

would represent games that were

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs (hunk 3 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: Now we tell user it's a coordinate descent method.


---
In reply to: [274574781](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274574781) [](ancestors = 274574781)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs (hunk 5 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: Thanks.

---
In reply to: [274575597](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274575597) [](ancestors = 274575597)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs (hunk 6 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: I'd like to be more explicit. :)

---
In reply to: [274575710](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274575710) [](ancestors = 274575710)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs (hunk 7 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: New description
```csharp
                // The desired values of matrix elements not specified in the training set.
                // If the training set doesn't tell the value at the u-th row and v-th column,
                // its desired value would be set 0.15. In other words, this parameter determines
                // the value of all missing matrix elements.
```

---
In reply to: [274575901](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274575901) [](ancestors = 274575901)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs (hunk 9 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Recommendation/OneClassMatrixFactorizationWithOptions.cs]: This is long but makes the code readable like a text message.

---
In reply to: [274576483](https://github.com/dotnet/machinelearning/pull/3282#discussion_r274576483) [](ancestors = 274576483)

**Pull request merged**